src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,308 @@
+                public List<Tuple<DelayedType, DelayedType>>? Others { get; set; }
+
+                public bool Add(Tuple<DelayedType, DelayedType> newEntryMaybe)
+                {
+                    Others ??= new List<Tuple<DelayedType, DelayedType>>();
+                    foreach (Tuple<DelayedType, DelayedType> entry in Others)
+                    {
+                        if (entry.Item1.RepresentsSameType(newEntryMaybe.Item1))
+                        {
+                            return false;
+                        }
+                    }
+
+                    Others.Add(newEntryMaybe);
+                    return true;
+                }
+            }
+
+            private readonly Dictionary<int, DelayedTypeCollection> _lazyData = new();
+
+            protected override bool TryGetOrLoadType(Type key, [NotNullWhen(true)] out Type? type)
+            {
+                int hash = ComputeHashCode(key.FullName!);
+
+                if (_lazyData.TryGetValue(hash, out DelayedTypeCollection? value))
+                {
+                    // The common case, no duplicate mappings.
+                    if (value.First.Item1.GetOrLoadType() == key)
+                    {
+                        type = value.First.Item2.GetOrLoadType();
+                        return true;
+                    }
+                    else if (value.Others != null)
+                    {
+                        // Common case failed, look at alternate mappings.
+                        foreach (Tuple<DelayedType, DelayedType> entry in value.Others)
+                        {
+                            if (entry.Item1.GetOrLoadType() == key)
+                            {
+                                type = entry.Item2.GetOrLoadType();
+                                return true;
+                            }
+                        }
+                    }
+                }
+                type = null;
+                return false;
+            }
+
+            public void Add(TypeName parsedSource, TypeName parsedProxy)
+            {
+                // We don't use the assembly name for the hash code since it is not
+                // guaranteed to be the same for the same type due to type forwarding.
+                int hash = ComputeHashCode(parsedSource.FullName);
+
+                Tuple<DelayedType, DelayedType> newEntryMaybe = new(
+                    new(parsedSource.AssemblyName!.FullName, parsedSource.FullName),
+                    new(parsedProxy.AssemblyName!.FullName, parsedProxy.FullName));
+
+                if (!_lazyData.TryGetValue(hash, out DelayedTypeCollection? types))
+                {
+                    types = new DelayedTypeCollection() { First = newEntryMaybe };
+                    _lazyData.Add(hash, types);
+                }
+                else
+                {
+                    if (!types.Add(newEntryMaybe))
```

Comment:
```suggestion
                    if (types.First.Item1.RepresentsSameType(newEntryMaybe.Item1)
                        || !types.Add(newEntryMaybe))
```

New to check the `First` entry too.